File: src/distribution/distribution_master.go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/go-redis/redis/v8"
	"github.com/sirupsen/logrus"
)

type DistributedMaster struct {
	port            int
	redisClient     *redis.Client
	resultHashKey   string
	taskDataHashKey string
	logger          *logrus.Logger
	taskRestCtx     context.Context
	taskRestCancel  context.CancelFunc
	wg              sync.WaitGroup
}

func NewDistributedMaster(port int) *DistributedMaster {
	// 初始化日志
	logger := logrus.New()
	logDir := filepath.Join("..", "..", "logs", "Distribution_logs")
	if err := os.MkdirAll(logDir, 0755); err != nil {
		log.Fatalf("Failed to create log directory: %v", err)
	}
	logFile := filepath.Join(logDir, "master.log")
	file, err := os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("Failed to open log file: %v", err)
	}
	logger.SetOutput(file)
	logger.SetLevel(logrus.InfoLevel)

	// 初始化Redis
	rdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		PoolSize: 256,
	})

	// 清理现有数据
	ctx := context.Background()
	rdb.Del(ctx, "distributed_result", "task_data_hash", "cpu", "gpu")

	// 创建上下文用于后台任务
	taskCtx, cancel := context.WithCancel(context.Background())

	return &DistributedMaster{
		port:            port,
		redisClient:     rdb,
		resultHashKey:   "distributed_result",
		taskDataHashKey: "task_data_hash",
		logger:          logger,
		taskRestCtx:     taskCtx,
		taskRestCancel:  cancel,
	}
}

func (dm *DistributedMaster) Run() {
	router := gin.Default()
	router.Use(dm.checkVerification())

	router.POST("/get_task", dm.getTask)
	router.POST("/get_result", dm.getResult)
	router.POST("/clear_record", dm.clearRecord)
	router.POST("/report_result", dm.reportResult)
	router.POST("/create_task", dm.createTask)

	// 启动后台任务
	dm.wg.Add(1)
	go dm.showTaskRest()

	// 启动HTTP服务器
	srv := &http.Server{
		Addr:    fmt.Sprintf(":%d", dm.port),
		Handler: router,
	}

	go func() {
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			dm.logger.Fatalf("Server error: %v", err)
		}
	}()

	// 等待中断信号
	// 	quit := make(chan os.Signal, 1)
	// 	signal.Notify(quit, os.Interrupt)
	// 	<-quit
	//
	// 	// 优雅关闭
	// 	dm.taskRestCancel()
	// 	dm.wg.Wait()
	//
	// 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	// 	defer cancel()
	// 	if err := srv.Shutdown(ctx); err != nil {
	// 		dm.logger.Fatal("Server shutdown error:", err)
	// 	}
}

func (dm *DistributedMaster) checkVerification() gin.HandlerFunc {
	return func(c *gin.Context) {
		checkVal := c.PostForm("check_val")
		if checkVal != "81600a92e8416bba7d9fada48e9402a4" {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"success": false,
				"msg":     "ERROR",
			})
			return
		}
		c.Next()
	}
}

func (dm *DistributedMaster) getTask(c *gin.Context) {
	taskType := c.PostForm("task_type")
	if taskType != "cpu" && taskType != "gpu" {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"msg":     "TASK Type ERROR",
		})
		return
	}

	maxCost, err := strconv.ParseFloat(c.PostForm("max_cost"), 64)
	if err != nil {
		dm.logger.Error("Invalid max_cost:", err)
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid max_cost"})
		return
	}

	var taskID string
	for {
		// ZRANGEBYSCORE with LIMIT 0 1
		cmd := dm.redisClient.ZRangeByScore(c.Request.Context(), taskType, &redis.ZRangeBy{
			Min:    "0",
			Max:    strconv.FormatFloat(maxCost, 'f', -1, 64),
			Offset: 0,
			Count:  1,
		})

		members, err := cmd.Result()
		if err != nil {
			dm.logger.Error("Redis error:", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		if len(members) == 0 {
			c.JSON(http.StatusInternalServerError, gin.H{
				"success": false,
				"msg":     "There is no Task can RUN",
			})
			return
		}

		// 使用事务确保原子性
		txPipe := dm.redisClient.TxPipeline()
		remCmd := txPipe.ZRem(c.Request.Context(), taskType, members)
		_, err = txPipe.Exec(c.Request.Context())
		if err != nil {
			dm.logger.Error("Transaction failed:", err, "Retry")
			continue
		}

		// 确保唯一取到taskID
		if remCmd.Val() > 0 {
			taskID = members[0]
			break
		}
	}

	//取出对应taskID的taskData
	result, err := dm.redisClient.HGet(c.Request.Context(), dm.taskDataHashKey, taskID).Result()
	if err == redis.Nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"msg":     "Task LOSS",
		})
		return
	} else if err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	// 删除taskID的taskData记录
	if _, err := dm.redisClient.HDel(c.Request.Context(), dm.taskDataHashKey, taskID).Result(); err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Data(http.StatusOK, "application/octet-stream", []byte(result))
}

func (dm *DistributedMaster) clearRecord(c *gin.Context) {
	taskID := c.PostForm("task_id")
	if _, err := dm.redisClient.HDel(c.Request.Context(), dm.resultHashKey, taskID).Result(); err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"success": true})
}

func (dm *DistributedMaster) getResult(c *gin.Context) {
	taskID := c.PostForm("task_id")
	result, err := dm.redisClient.HGet(c.Request.Context(), dm.resultHashKey, taskID).Result()
	if err == redis.Nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"msg":     "Task haven't FINISHED",
		})
		return
	} else if err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Data(http.StatusOK, "application/octet-stream", []byte(result))
}

func (dm *DistributedMaster) reportResult(c *gin.Context) {
	taskID := c.PostForm("task_id")
	file, err := c.FormFile("result")
	if err != nil {
		dm.logger.Error("File error:", err)
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	f, err := file.Open()
	if err != nil {
		dm.logger.Error("File open error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer f.Close()

	buf := make([]byte, file.Size)
	_, err = f.Read(buf)
	if err != nil {
		dm.logger.Error("File read error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if _, err := dm.redisClient.HSet(c.Request.Context(), dm.resultHashKey, taskID, buf).Result(); err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "msg": "SUCCESS"})
}

func (dm *DistributedMaster) createTask(c *gin.Context) {
	file, err := c.FormFile("task_data")
	if err != nil {
		dm.logger.Error("File error:", err)
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	taskType := c.PostForm("task_type")
	if taskType != "cpu" && taskType != "gpu" {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"msg":     "TASK Type ERROR",
		})
		return
	}

	taskID := c.PostForm("task_id")

	taskCost, err := strconv.ParseFloat(c.PostForm("task_cost"), 64)
	if err != nil {
		dm.logger.Error("Invalid task_cost:", err)
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid task_cost"})
		return
	}

	f, err := file.Open()
	if err != nil {
		dm.logger.Error("File open error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer f.Close()

	buf := make([]byte, file.Size)
	_, err = f.Read(buf)
	if err != nil {
		dm.logger.Error("File read error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if _, err := dm.redisClient.HSet(c.Request.Context(), dm.taskDataHashKey, taskID, buf).Result(); err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if _, err := dm.redisClient.ZAdd(c.Request.Context(), taskType, &redis.Z{
		Score:  taskCost,
		Member: taskID,
	}).Result(); err != nil {
		dm.logger.Error("Redis error:", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true, "msg": "SUCCESS"})
}

func (dm *DistributedMaster) showTaskRest() {
	defer dm.wg.Done()
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			ctx := context.Background()
			cpuCount, _ := dm.redisClient.ZCard(ctx, "cpu").Result()
			gpuCount, _ := dm.redisClient.ZCard(ctx, "gpu").Result()
			dm.logger.Infof("REST TASK NUM is %d", cpuCount+gpuCount)
		case <-dm.taskRestCtx.Done():
			return
		}
	}
}

func main() {
	port := 1088
	master := NewDistributedMaster(port)
	master.Run()
	for true {
		time.Sleep(1 * time.Second)
	}
}

File: src/distribution/distribution_master.py
import logging
import os
import time
from multiprocessing import Manager, Process, Lock
from pathlib import Path

import logzero
import redis
import urllib3
from flask import Flask, request, make_response


urllib3.disable_warnings(category=urllib3.exceptions.InsecureRequestWarning)


class DistributedMaster(Process):
    def __init__(self, port: int = 1088):
        super().__init__()
        if not os.path.exists(Path(os.path.dirname(os.path.abspath(__file__)), "../../logs/Distribution_logs")):
            os.makedirs(Path(os.path.dirname(os.path.abspath(__file__)), "../../logs/Distribution_logs"))
        self.logger = logzero.setup_logger(
            logfile=str(Path(os.path.dirname(os.path.abspath(__file__)), "../../logs/Distribution_logs/master.log")),
            name="DistributedMaster Log", level=logzero.ERROR, fileLoglevel=logzero.INFO, backupCount=100,
            maxBytes=int(1e7))
        self.port = port
        self.manager = Manager()
        self.redis_pool = redis.ConnectionPool(host='localhost', port=6379, db=0, max_connections=256)
        self.redis = redis.Redis(connection_pool=self.redis_pool)
        self.result_hash_key = "distributed_result"
        self.redis.delete(self.result_hash_key)
        self.redis.delete("cpu")
        self.redis.delete("gpu")
        self.task_lock = Lock()
        self.app = Flask(__name__)
        self.app.add_url_rule("/get_task", view_func=self.get_task, methods=['POST'])
        self.app.add_url_rule("/get_result", view_func=self.get_result, methods=['POST'])
        self.app.add_url_rule("/clear_record", view_func=self.clear_record, methods=['POST'])
        self.app.add_url_rule("/report_result", view_func=self.report_result, methods=['POST'])
        self.app.add_url_rule("/create_task", view_func=self.create_task, methods=['POST'])
        werkzeug_logger = logging.getLogger('werkzeug')
        werkzeug_logger.disabled = True
        self.set_middleware()
        self.task_num_process = Process(target=self.show_task_rest, args=())
        self.task_num_process.start()

    def run(self):
        self.app.run(debug=False, port=self.port, host='0.0.0.0', threaded=False, processes=64)

    def set_middleware(self):
        @self.app.before_request
        def check_verification():
            try:
                data = request.form.to_dict()
                check_val = data["check_val"]
                if check_val != "81600a92e8416bba7d9fada48e9402a4":
                    return {"success": False, "msg": "ERROR"}
            except Exception as e:
                self.logger.error(str(e))
                return {"success": False, "msg": str(e)}

    def get_task(self):
        try:
            data = request.form.to_dict()
            task_type = data["task_type"]
            if task_type not in ["cpu", "gpu"]:
                return make_response({"success": False, "msg": "TASK Type ERROR"}, 500)
            max_cost = int(data["max_cost"])
            rem_num = 0
            while rem_num <= 0:
                results = self.redis.zrangebyscore(
                    task_type, 0, max_cost,
                    start=0, num=1, withscores=False
                )
                if results:
                    rem_num = self.redis.zrem(task_type, results[0])
                else:
                    return make_response({"success": False, "msg": "There is no Task can RUN"}, 500)
            pickled_data = results[0]
            response = make_response(pickled_data, 200)
            response.headers['Content-Type'] = 'application/octet-stream'
            return response
        except Exception as e:
            self.logger.error("There is ERROR: " + str(e))
            return make_response({"success": False, "msg": str(e)}, 500)

    def clear_record(self):
        try:
            data = request.form.to_dict()
            task_id = data["task_id"]
            self.redis.hdel(self.result_hash_key, task_id)
        except Exception as e:
            self.logger.error(str(e))
            return make_response({"success": False, "msg": str(e)}, 500)
        return make_response({"success": True}, 200)

    def get_result(self):
        try:
            data = request.form.to_dict()
            task_id = data["task_id"]
            result = self.redis.hget(self.result_hash_key, task_id)
            # if result is None:
            #     return make_response({"success": False, "msg": "Task does not EXIST"}, 500)
            if result is None:
                return make_response({"success": False, "msg": "Task haven't FINISHED"}, 500)
            pickled_data = result
            response = make_response(pickled_data, 200)
            response.headers['Content-Type'] = 'application/octet-stream'
            return response
        except Exception as e:
            self.logger.error(str(e))
            return make_response({"success": False, "msg": str(e)}, 500)

    def report_result(self):
        try:
            data = request.form.to_dict()
            task_id = data["task_id"]
            task_result = request.files['result'].stream.read()
            self.redis.hset(self.result_hash_key, task_id, task_result)
            return make_response({"success": True, "msg": "SUCCESS"}, 200)
        except Exception as e:
            self.logger.error(str(e))
            return make_response({"success": False, "msg": str(e)}, 500)

    def create_task(self):
        try:
            task_data = request.files['task_data'].stream.read()
            data = request.form.to_dict()
            try:
                task_type = data["task_type"]
                if task_type not in ["cpu", "gpu"]:
                    return make_response({"success": False, "msg": "TASK Type ERROR"}, 500)
                task_cost = data["task_cost"]
                self.redis.zadd(task_type, {task_data: task_cost})
                return make_response({"success": True, "msg": "SUCCESS"}, 200)
            except KeyError:
                return make_response({"success": False, "msg": "TASK Param Missing"}, 500)
        except Exception as e:
            self.logger.error(str(e))
            return make_response({"success": False, "msg": str(e)}, 500)

    def show_task_rest(self):
        while True:
            time.sleep(5)
            task_rest = self.redis.zcard("cpu") + self.redis.zcard("gpu")
            self.logger.info(f"REST TASK NUM is {task_rest}")


if __name__ == '__main__':
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument('--port', type=int, default=1088, help='The port of the Master')
    args = parser.parse_args()
    master = DistributedMaster(port=args.port)
    master.start()
    master.join()
